### ISSUE TITLE
getifaddrs() with ns3 stack does not seem to answer

### ISSUE THREAD

**Issue by teto** [direct-code-execution/ns-3-dce#2]
This is a follow up of https://groups.google.com/forum/#!topic/ns-3-users/EoZ6DssRiqM where a call to getifaddr seemingly returns no result.
I was wondering if DCE configured with a ns3 stack could answer a query for RTM_GETLINK ? 
I've seen that line in the code so I am not the first to see it ^^:
```// msg->msg_controllen = 0; // why???```
Is there any development or procedure needed to get this work? I can help on this if dev is needed

**Comment by teto** [comment]
My understanding is that dce_getifaddr generates a ntlink request and send it to an ns3 socket that does not handle netlink, thus returning a null message. Also in the code, I see in comments before dce_ifgetaddr:
```
/*
 * Try to emulate netlink socket query to work both ns3 stack and
 * linux stack.
 */
```
My solution/idea is to detect when it's a ns3 stack and act differently but I am not sure about scalability of such a solution
```
int
dce_getifaddrs (struct ifaddrs **ifap)
{
  NS_LOG_INFO("MATT dce_getifaddr");
//Depending on the socket factory type Id I can distinguish if I run a real or ns3 stack.
// TODO move the following code to an enum UtilsGetStackType()
// we cna't do ns3::Ns3SocketFdFactory::GetTypeId() since the header is not exported according to the wscript

  Ptr<Node> node = UtilsGetNode();

  TypeId t = TypeId::LookupByName ("ns3::Ns3SocketFdFactory");
  Ptr<SocketFdFactory> facto = node->GetObject<SocketFdFactory>( t );
  if(facto){
    return dce_getifaddrs_ns3(ifap);
  }
...
```

**Comment by thehajime** [comment]
I think switching the behavior of dce_getifaddrs() and using dce_getifaddrs_ns3() are not a problematic.
Since DCE only has a few stacks (Ns3SocketFdFactory, Linux, and FreeBSD: latter two are almost similar), it's not likely to be an issue with these line.

btw, dce_getifaddrs() originally intended to support both stack (ns3 and Linux): ns3::NetlinkSocket should work with that.

Since netlink with ns3 stack is not well tested (with linux stack and netlink, we've been tested pretty much), I think there is some issue in our implementation of netlink wrapper (i.e., ns-3-dce/netlink). I glanced at the behavior of getifaddr with ns-3 stack but not sure what's going on.


so if you really need ns3 stack and wanna use getifaddr(), I would suggest to fix the ns3::NetlinkSocket issue to work correctly with dce_getifaddr(). having a workaround with dce_getifaddrs_ns3() would be also nice, but it hides other issues of ns3::NetlinkSocket, which you may face later on.

**Comment by teto** [comment]
I see your point. I would like to be able to export ns3 netlink communications to a pcap file for easier debugging. What would be the best place to do so in your opinion ? I started creating a NetlinkNetDevice but it seems overkill.

**Comment by thehajime** [comment]
I believe ns3::NetlinkSocket is the appropriate place to export pcap.

**Comment by teto** [comment]
I am able to export packets sent by ns3 by adding a TraceSource in NetlinkSocket::ForwardUp but I can't seem to see packets sent by the application (for instance ntpd) through the use of dce_getifaddr. dce_getifaddr  calls netlink_request. I guess I should add some tracesource here too. I will upload a version soon if that can help.

**Comment by teto** [comment]
I put a tracesource in NetlinkSocket:SendTo  and apparently I am able to catch everything :)
Now I have to understand why dce_getifaddr does not work with ntpd

**Comment by thehajime** [comment]
great.

**Comment by teto** [comment]
I am closing in on the problem, the netlink packet is apparently received the msg_iovlen is <0 while it  should not be the case and thus gets discarded.
I've added logging/comments to Cmsg.h . I noticed a possible problem in the API, you have no way of knowing if adding ancillary data failed or not (  if (m_len < cmsglen) ):
```
void
Cmsg::Add (int level, int type, int len, const uint8_t *buffer)
{
  NS_LOG_DEBUG(level << type << len << buffer);
  int cmsglen = CMSG_SPACE (len);
  if (m_len < cmsglen)
    {
      NS_LOG_WARN("Can't append message: not enough space");
      m_msg->msg_flags |= MSG_CTRUNC;
      return;
    }

```
Would it be ok to change the API in a future PR to return a boolean ?

**Comment by thehajime** [comment]
yes, you can do any PR, especially Cmsg::Add is now internally used within DCE so, it might not be an issue for the API change.

**Comment by teto** [comment]
I was wrong: the problem had nothing to do with Cmsg. When comparing real netlink traces with ns3 traces, I noticed the port id pattern was different: dce was using it to put source address while it should be the destination address. Also DCE was always setting the seq number to 0 (so the first call of dce_getifaddr would work but not the second for instance). 
I've fixed the problem in the places where I was interested but it may require another pass or a more holistic solutions.
I also had ntpd complaining that IFA_NAME was not set to answer GETADDR. I fixed it too.
I now have another problem (didn't want to create a new ticket), i.e. dce_if_nametoindex checks to see if the factory is a linux one:
```  if (factory->GetInstanceTypeId () == TypeId::LookupByName ("ns3::LinuxSocketFdFactory")) ```
but ns3 asserts because the typeId "ns3::LinuxSocketFdFactory" was never registered to begin with. I believe it is because waf script needs    `if bld.env['KERNEL_STACK']:` to embed LinuxSocketFdFactory. Not having KERNEL_STACK prevents DCE from being used with ns3 alone which shouldn't be. How do you propose to fix the wscript so that "ns3::LinuxSocketFdFactory" gets registered even without 'KERNEL_STACK' ? I tried moving the line 'model/linux-socket-fd-factory.cc' in another place of wscript but the dependancies generate errors etc...

Here is a backtrace linked to the error
```
56	../nptl/sysdeps/unix/sysv/linux/raise.c: Aucun fichier ou dossier de ce type.
(gdb) bt
#0  0x00007fffee3cde37 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007fffee3cf528 in __GI_abort () at abort.c:89
#2  0x00007fffeebf105d in __gnu_cxx::__verbose_terminate_handler() ()
   from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007fffeebeeed6 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007fffeebeef21 in std::terminate() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff73cd32b in ns3::TypeId::LookupByName (name="ns3::LinuxSocketFdFactory")
    at ../src/core/model/type-id.cc:560
#6  0x00007ffff7a7a8b1 in dce_if_nametoindex (ifname=0x7ffff7fc62c8 "ns3-device1")
    at ../model/dce.cc:684
#7  0x00007fffea372330 in if_nametoindex () at ../model/libc-ns3.h:501
#8  0x00007fffe91a2936 in internal_current (iter=0x7ffff7fc6298)
    at ./../lib/isc/unix/ifiter_getifaddrs.c:237
#9  0x00007fffe91a2f22 in isc_interfaceiter_first (iter=0x7ffff7fc6298)
    at ./../lib/isc/unix/interfaceiter.c:282
#10 0x00007fffe91281fc in update_interfaces (port=123, receiver=0x0, data=0x0) at ntp_io.c:1781
#11 0x00007fffe9128ac6 in create_sockets (port=123) at ntp_io.c:2068
#12 0x00007fffe9125745 in io_open_sockets () at ntp_io.c:493
#13 0x00007fffe91247ae in config_ntpd (ptree=0x7ffff7fed5e8, input_from_files=1) at ntp_config.c:4314
#14 0x00007fffe9124c96 in save_and_apply_config_tree (input_from_file=1) at ntp_config.c:4546
#15 0x00007fffe9124ab7 in getconfig (argc=0, argv=0x695d60) at ntp_config.c:4482
#16 0x00007fffe9132e6f in ntpdmain (argc=0, argv=0x695d60) at ntpd.c:792
#17 0x00007fffe91323d7 in main (argc=4, argv=0x695d40) at ntpd.c:289
#18 0x00007ffff7a5180d in ns3::DceManager::DoStartProcess (context=0x695f30)
    at ../model/dce-manager.cc:283
```
Waiting for your input.

**Comment by thehajime** [comment]
At Wed, 10 Jun 2015 08:59:30 -0700,
Matthieu Coudron wrote:

> I now have another problem (didn't want to create a new ticket), i.e. dce_if_nametoindex checks to see if the factory is a linux one:
> ```  if (factory->GetInstanceTypeId () == TypeId::LookupByName ("ns3::LinuxSocketFdFactory")) ```

how about 's/LookupByName/LookupByNameFailSafe/' ?

# I would prefer to have a different issue but it's okay :)
 
-- Hajime

**Comment by teto** [comment]
Great idea, didn't know the Failsafe version thanks.
I agree that creating a different issue would have been better but I hesitate to create more issues as I don't know how you consider github compared to the ns3 bugzilla. I personnally enjoy the github workflow.

**Comment by thehajime** [comment]
for the current status of github operation with DCE, I'm now in a migration phase: I'm thinking to converge  to github-based development in near future while existing ones (bugzilla, hg repos, etc) are also maintained.

so feel free to choose your preferred way: I think I'm even ready to accept PR on github.

happy hacking !

**Comment by teto** [comment]
I've not forgotten this, I hit a problem within ns3 callbacks, waiting to see @pdbj solution. Hopefully I should be able to send a PR soon

**Comment by thehajime** [comment]
@teto not sure what problem you faced but i'm happy to have any PR.

**Comment by teto** [comment]
What I have in my fork and plan to upstream (in several PRs maybe):
- upstream DCE netlink seem to invert source and destination address, got a fix for that
- possibility to export netlink (that's the blocking part, depends on ns3 changes)
- a few syscalls to let ntpd run within DCE

**Comment by thehajime** [comment]
fixed by #23. closed.

(reopen it if it's not right)
